text/0026-block-version-based-protocol-evolution.md

```diff
@@ -0,0 +1,288 @@
+- Feature Name: Block-Version-based Protocol Evolution
+- Start Date: 2022-02-05
+- MCIP PR: [mobilecoinfoundation/mcips#0026](https://github.com/mobilecoinfoundation/mcips/pull/0026)
+- MobileCoin Epic: None
+
+# Summary
+[summary]: #summary
+
+Since inception, MobileCoin has not yet made any changes to the ledger format.
+
+This document proposes a general approach to rolling out changes to the ledger format
+efficiently while carefully maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+Conventional web applications today are usually developed using data interchange
+formats that support schema evolution with forward and backwards compatibility.
+A good example of this is protobuf -- protobuf allows that new fields can be
+added to schema, and new programs will be able to deserialize old payloads, and
+old programs will be able to deserialize the new payloads (ignoring the new fields).
+
+As long as changes are strictly additive then they can be rolled out without
+breaking old clients. Eventually the old clients are sunsetted and only the new
+clients are supported.
+
+In blockchains, this is made more complicated for two reasons:
+* Older data payloads exist forever in the blockchain and clients have to be able to handle them forever
+* Frequently, clients and nodes need to agree about the hash of of a piece of data, so that
+  clients that sync the blockchain compute the correct block id, and clients that submit merkle proofs
+  of membership compute correct proofs.
+
+Because of the need for everyone to compute hashes correctly, it may not be enough for
+an old client to successfully deserialize a new payload by ignoring the fields it doesn't
+understand -- if those fields are ignored, then it will compute a different hash for
+this object from the new clients.
+
+To make this concrete, this means that if we want to roll out changes that add new
+features to `TxOut`'s, the new clients and old clients probably cannot coexist the way
+they might in a more conventional application.
+
+If a new client sends a `TxOut` to a user using an old client, which contains a `TxOut`
+feature that they don't recognize, then:
+1. They may fail to compute the correct block id
+2. If they submit a transaction spending the TxOut, the merkle proof of membership
+   may be wrong if the hash doesn't incorporate the new field.
+3. Even if they are a fog client, and (new) fog is building the proof of membership for them,
+   if the client strips out the fields it doesn't understand that fog gives them, then consensus
+   will likely reject the transaction. (There are some protobuf implementations that support "preserving"
+   unknown fields, but PROST, which is the only rust implementation which is enclave-compatible, does not do this.)
```

Comment:
I wonder if it makes sense for us to invest in making Prost support preserving unknown fields.